### PR TITLE
Skip dynamic render path check for renderables

### PR DESCRIPTION
**Pull request by bradparker**
[Rails 6.1 now supports passing `Renderable`s to `render`](https://github.com/rails/rails/blob/b5b078cd99b0107fd8e31ff619578e5321182e0a/actionview/lib/action_view/template/renderable.rb).

A renderable is anything that responds to `render_in`. This means we can
now expect to see `render(SomeRenderableClass.new(some_parameter))`. A
new library for building Rails UIs out of components, ViewComponent, is
growing in popularity and makes use of this new `render_in` API.

Closes https://github.com/presidentbeef/brakeman/issues/1529.

### CODE REVIEW

File: lib/brakeman/checks/check_render.rb [hunk 1]
```diff
@@ -94,4 +95,21 @@ def safe_param? exp
...
-end 
+
+  def renderable? exp
+    call?(exp) and
+      sexp?(exp.target) and
+      constant?(exp.target) and
+      method_defined?(class_name(exp.target), :render_in)
+  end
+
+  def method_defined? class_name, name, seen = []
```
Comment (bradparker) [inline lib/brakeman/checks/check_render.rb:106]: I'm not sure about where best to put this, it feels too general too be scoped to only checking render calls. I also wanted to search `#includes` for `render_in` but the collection was always empty for my test components. I could see includes being populated for models but not for libs ... I could look better into how to populate includes? However, without that I _think_ this should address the ViewComponent use case as it looks like `render_in` is defined on `ViewComponent::Base`.

**Comment by presidentbeef** [comment]
Hi @bradparker! Wow thanks for taking a swing at this! 🙏 

I think I have some pending changes that will help with method lookup.

I can't look too deeply at this PR right now but will follow up in the next few days!

**Comment by bradparker** [comment]
> I think I have some pending changes that will help with method lookup.

Rad, thanks.

> I can't look too deeply at this PR right now but will follow up in the next few days!

No problem :). I feel like I'm probably missing some things so I really appreciate you taking the time, when ever you might have it, to have a look.

**Comment by bradparker** [comment]
Ah, right. So, silly me, I've checked this in our code base just now. I didn't realize that Brakeman doesn't index dependencies so this doesn't quite work. It looks up `ViewComponent::Base` but can't find it so doesn't identify ViewComponents as renderables ... perhaps it would be possible to configure some class names as "renderables" if you're not comfortable adding specific handling for ViewComponent?

**Comment by bradparker** [comment]
https://github.com/presidentbeef/brakeman/pull/1578/commits/8fe395edeb000b6f170f7f7093d31c392f41b7cf is a perhaps a proof of concept for handling ViewComponent specially. I've plugged that into our code base and it's removed all the ViewComponent related 'dynamic render path' false positives.

File: lib/brakeman/checks/check_render.rb [hunk 2]
```diff
@@ -94,4 +95,37 @@ def safe_param? exp
...
+      (known_renderable_class?(class_name(exp.target)) or method_defined?(class_name(exp.target), :render_in))
+  end
+
+  def known_renderable_class? start
+    traverse_ancestry(start) do |class_name|
+      return true if class_name == :"ViewComponent::Base"
+    end
+  end
+
+  def method_defined? start, name
```
Comment (presidentbeef) [inline lib/brakeman/checks/check_render.rb:112]: Check out `Tracker#find_method` I just merged from https://github.com/presidentbeef/brakeman/pull/1596 I think it takes care of this.

File: lib/brakeman/checks/check_render.rb [hunk 3]
```diff
@@ -94,4 +95,37 @@ def safe_param? exp
...
+  def renderable? exp
+    call?(exp) and
+      sexp?(exp.target) and
+      constant?(exp.target) and
+      (known_renderable_class?(class_name(exp.target)) or method_defined?(class_name(exp.target), :render_in))
+  end
+
+  def known_renderable_class? start
+    traverse_ancestry(start) do |class_name|
+      return true if class_name == :"ViewComponent::Base"
```
Comment (presidentbeef) [inline lib/brakeman/checks/check_render.rb:108]: I think you can use `Collection#ancestor?` for this:

```ruby
klass = tracker.find_class(class_name)
klass.ancestor? :"ViewComponent::Base"
```

**Comment by presidentbeef** [comment]
> [8fe395e](https://github.com/presidentbeef/brakeman/commit/8fe395edeb000b6f170f7f7093d31c392f41b7cf) is a perhaps a proof of concept for handling ViewComponent specially. I've plugged that into our code base and it's removed all the ViewComponent related 'dynamic render path' false positives.

That seems totally reasonable to me.

File: lib/brakeman/checks/check_render.rb [hunk 4]
```diff
@@ -94,4 +95,17 @@ def safe_param? exp
       end
     end
   end
-end 
+
+  def renderable? exp
+    return false unless call?(exp) and sexp?(exp.target) and constant?(exp.target)
```
Comment (presidentbeef) [inline lib/brakeman/checks/check_render.rb:100]: `constant?` will check if it's a `Sexp` so you can skip the `sexp?` call.

File: lib/brakeman/checks/check_render.rb [hunk 5]
```diff
@@ -94,4 +95,17 @@ def safe_param? exp
...
+
+    target_class_name = class_name(exp.target)
+    known_renderable_class?(target_class_name) or tracker.find_method(:render_in, target_class_name)
+  end
+
+  def known_renderable_class? class_name
```
Comment (presidentbeef) [inline lib/brakeman/checks/check_render.rb:106]: Can you add a test for this? Thanks!

**Review by presidentbeef — CHANGES_REQUESTED** [review]

**Comment by presidentbeef** [comment]
Thanks @bradparker!

**Comment by bradparker** [comment]
No worries! Thanks :)

**Pull request merged**